### PR TITLE
Add an option to delete versions of a crate

### PR DESCRIPTION
**Pull request by jyn514**
Closes https://github.com/rust-lang/docs.rs/issues/528.
Allows closing https://github.com/rust-lang/docs.rs/issues/837 and https://github.com/rust-lang/docs.rs/issues/158.
Makes https://github.com/rust-lang/docs.rs/issues/368 much easier.
Makes https://github.com/rust-lang/docs.rs/issues/766 more useful.

r? @pietroalbini

### CODE REVIEW

**Comment by jyn514** [comment]
This is best reviewed commit-by-commit since I renamed `delete_crates.rs`.

File: src/db/delete_crate.rs [hunk 1]
```diff
@@ -14,22 +14,89 @@ enum CrateDeletionError {
...
+    let crate_id_res = conn.query("SELECT id FROM crates WHERE name = $1", &[&name])?;
+    if let Some(row) = crate_id_res.into_iter().next() {
+        Ok(row.get("id"))
+    } else {
+        Err(CrateDeletionError::MissingCrate(name.into()).into())
     }
+}
 
+// metaprogramming!
+// NOTE: it is _crucial_ that table_name and column_name be hard-coded, i.e. NOT user input
```
Comment (pietroalbini) [inline src/db/delete_crate.rs]: This makes me a bit worried.

The `delete_metadata` function is only used in `delete_version_from_database`: we could inline it, which would remove the risk of other code miscalling it.

Comment (jyn514) [reply src/db/delete_crate.rs]: Good idea, done.

**Comment by pietroalbini** [comment]
Feel free to merge once CI passes.

**Pull request merged**